### ISSUE TITLE
Bug: Cloud upload not correct?

### ISSUE THREAD

**Issue by fs1209** [TortugaPower/BookPlayer#1081]
## Environment

 - OS Version: iPadOS / iOS 17.3
 - App Version: 5.1.1-28

## Description

The cloud upload for some of the files are (most probably) not correct.

One example (see screenshots attached) the file "04 Goblet of Fire" has 719.4 MB and was according to the app uploaded to the cloud. On any other device (I've tried it with another iPhone, an iPad and a Mac) the file has 327 Bytes.
These files show no cover on the new device.

## Steps to reproduce the behavior
1. Install BookPlayer on a device
2. Log into the account via Apple ID
3. Download the file

iPhone folder BookPlayer/Processed:
![image](https://github.com/TortugaPower/BookPlayer/assets/150681522/1f64b80b-2bbc-481a-9fad-f2648bba1f29)

iPad:
![image](https://github.com/TortugaPower/BookPlayer/assets/150681522/7f489edb-bc95-4081-8553-7fa5c1eba056)

**Comment by GianniCarlo** [comment]
Hi @fs1209 ! Can you share to my email gcarlo89@hotmail.com the debug information from your phone? I'll be able to verify more quickly what we have on our servers. I think I know what the bytes on the other devices are (probably an encoded AWS S3 error).

For additional context, I found an error (really a wrong assumption on my part) regarding how a third party library that we use to handle the sync tasks order worked. In short, between app launches, the sync tasks are not maintaining the original order, which can mess up replicating the actions taken in the app and the final hierarchy representation we have on the server. I already have our own implementation to remove the third party library, I just need to test it a little bit more before creating the PR and updating the Beta on TestFlight before release

Sorry about the issue, and thanks for your support to the app 🙌, I'll be waiting on the debug file for this ticket, to help solve anything that comes up

**Comment by fs1209** [comment]
Library
.
`-- [✓][☑] Audiobooks
|   |   `-- [✓][☑] 1984.m4a
|   |   `-- [✓][☐] A-Gruppe
|   |   |   |   `-- [✓][☐] 01 Misterioso (1).m4a
|   |   |   |   |-- [✓][☐] 02 Böses Blut (2).m4a
|   |   |   |   |-- [✓][☐] 03 Falsche Opfer (3).m4a
|   |   |   |   |-- [✓][☐] 04 Tiefer Schmerz (4).m4a
|   |   |   |   |-- [✓][☐] 05 Rosenrot (5).m4a
|   |   |   |   |-- [✓][☐] 06 Ungeschoren (6).m4a
|   |   |   |   |-- [✓][☐] 07 Totenmesse (7).m4a
|   |   |   |   |-- [✓][☐] 08 Dunkelziffer (8).m4a
|   |   `-- [✓][☑] Assassini.m4a
|   |   |-- [✓][☑] Ausgebrannt.m4a
|   |   `-- [✓][☑] Bartimäus
|   |   |   |   `-- [✓][☑] 01 Bartimäus - Das Amulett von Samarkand.m4a
|   |   |   |   |-- [✓][☑] 02 Bartimäus - Das Auge des Golem.m4a
|   |   |   |   |-- [✓][☑] 03 Bartimäus - Die Pforte des Magiers.m4a
|   |   |   |   |-- [✓][☑] 04 Bartimäus - Der Ring des Salomo.m4a
|   |   `-- [✓][☑] Battlefield Earth.m4a
|   |   |-- [✓][☑] Bis ich dich finde.m4a
|   |   |-- [✓][☑] Blackout.m4a
|   |   |-- [✓][☑] Carrie.m4a
|   |   |-- [✓][☑] Das Echo der Schuld.m4a
|   |   |-- [✓][☑] Das Herz der Hölle.m4a
|   |   |-- [✓][☑] Das Joshua-Profil.m4a
|   |   |-- [✓][☑] Das Moor des Vergessens.m4a
|   |   |-- [✓][☑] Das Opfer.m4a
|   |   |-- [✓][☑] Das Robbins Power Prinzip.mp4
|   |   |-- [✓][☑] Das schwarze Blut.m4a
|   |   |-- [✓][☑] Das Silmarillion.m4a
|   |   `-- [✓][☑] David Hunter
|   |   |   |   `-- [✓][☑] 01 Die Chemie des Todes.m4a
|   |   |   |   |-- [✓][☑] 02 Kalte Asche.m4a
|   |   `-- [✓][☑] Der Flug der Störche.m4a
|   |   |-- [✓][☑] Der Fönig.m4a
|   |   |-- [✓][☑] Der Glückspilz.m4a
|   |   `-- [✓][☑] Der Herr der Ringe
|   |   |   |   `-- [✓][☑] 01 Der Herr der Ringe - Die Gefährten.m4a
|   |   |   |   |-- [✓][☑] 02 Der Herr der Ringe - Die zwei Türme.m4a
|   |   |   |   |-- [✓][☑] 03 Der Herr der Ringe - Die Wiederkehr des Königs.m4a
|   |   `-- [✓][☑] Der Himmel auf Erden.m4a
|   |   |-- [✓][☑] Der Hobbit.m4a
|   |   |-- [✓][☑] Der Katalane.m4a
|   |   |-- [✓][☑] Der Medicus.m4a
|   |   |-- [✓][☑] Der Patient.m4a
|   |   |-- [✓][☑] Der Schwarm.m4a
|   |   |-- [✓][☑] Der Wald ist Schweigen.m4a
|   |   |-- [✓][☑] Diabolus.m4a
|   |   |-- [✓][☑] Die Entscheidung liegt bei dir.m4a
|   |   |-- [✓][☑] Die Kinder von Eden.m4a
|   |   |-- [✓][☑] Die Leopardin.m4a
|   |   |-- [✓][☑] Die Nadel.m4a
|   |   |-- [✓][☑] Die Pfeiler der Macht.m4a
|   |   |-- [✓][☑] Die purpurnen Flüsse.m4a
|   |   |-- [✓][☑] Die Rache.m4a
|   |   |-- [✓][☐] Die Siedler von Catan.m4a
|   |   |-- [✓][☐] Die Zucht.m4a
|   |   `-- [✓][☐] Dirk Gently
|   |   |   |   `-- [✓][☐] 01 Der elektrische Mönch - Dirk Gently 1.m4a
|   |   |   |   |-- [✓][☐] 02 Der lange dunkle Fünfuhrtee der Seele - Dirk Gently 2.m4a
|   |   |   |   |-- [✓][☐] 03 Lachs im Zweifel - Dirk Gently 3.m4a
|   |   `-- [✓][☑] DNA.m4a
|   |   `-- [✓][☐] Dora Gudmundsdottir
|   |   |   |   `-- [✓][☐] 01 Das letzte Ritual.m4a
|   |   |   |   |-- [✓][☐] 02 Das gefrorene Licht.m4a
|   |   |   |   |-- [✓][☐] 03 Das glühende Grab.m4a
|   |   |   |   |-- [✓][☐] 04 Die eisblaue Spur.m4a
|   |   |   |   |-- [✓][☐] 05 Feuernacht.m4a
|   |   |   |   |-- [✓][☐] 06 Todesschiff.m4a
|   |   `-- [✓][☐] Eberhofer
|   |   |   |   `-- [✓][☐] 01 Winterkartoffelknödel (1).m4a
|   |   |   |   |-- [✓][☐] 02 Dampfnudelblues (2).m4a
|   |   |   |   |-- [✓][☐] 03 Schweinskopf al dente (3).m4a
|   |   |   |   |-- [✓][☐] 04 Grießnockerlaffäre (4).m4a
|   |   |   |   |-- [✓][☐] 05 Sauerkrautkoma (5).m4a
|   |   |   |   |-- [✓][☐] 06 Zwetschgendatschikomplott (6).m4a
|   |   |   |   |-- [✓][☐] 07 Leberkäsjunkie (7).m4a
|   |   |   |   |-- [✓][☐] 08 Weißwurstconnection (8).m4a
|   |   |   |   |-- [✓][☐] 09 Kaiserschmarrndrama (9).m4a
|   |   |   |   |-- [✓][☐] 10 Guglhupfgeschwader (10).m4a
|   |   |   |   |-- [✓][☐] 11 Rehragout-Rendezvous (11).m4a
|   |   `-- [✓][☐] Eisfieber.m4a
|   |   |-- [✓][☐] Flieh in die dunkle Nacht.m4a
|   |   |-- [✓][☐] Geisterfjord.m4a
|   |   `-- [✓][☑] Glashaus
|   |   |   |   `-- [✓][☑] 01 Glashaus 1. Staffel.m4a
|   |   |   |   |-- [✓][☑] 02 Glashaus 2. Staffel.m4a
|   |   `-- [✓][☑] Harry Potter
|   |   |   |   `-- [✓][☑] 01 Philosopher's Stone.m4a
|   |   |   |   |-- [✓][☑] 02 Chamber of Secrets.m4a
|   |   |   |   |-- [✓][☑] 03 Prisoner of Azkaban.m4a
|   |   |   |   |-- [✓][☑] 04 Goblet of Fire.m4a
|   |   |   |   |-- [✓][☑] 05 Order of the Phoenix.m4a
|   |   |   |   |-- [✓][☑] 06 Half-Blood Prince.m4a
|   |   |   |   |-- [✓][☑] 07 Deathly Hallows.m4a
|   |   `-- [✓][☑] Helix.m4a
|   |   `-- [✓][☐] Hitchhiker's Guide to the Galaxy
|   |   |   |   `-- [✓][☐] 01 Hitchhiker's Guide to the Galaxy (1).m4a
|   |   |   |   |-- [✓][☐] 02 The Restaurant at the End of the Universe (2).m4a
|   |   |   |   |-- [✓][☐] 03 Life the Universe and Everything (3).m4a
|   |   |   |   |-- [✓][☐] 04 So Long and Thanks for All the Fish (4).m4a
|   |   |   |   |-- [✓][☐] 05 Mostly Harmless (5).m4a
|   |   `-- [✓][☐] Im Wald der stummen Schreie.m4a
|   |   |-- [✓][☐] In Sachen Signora Brunetti.m4a
|   |   |-- [✓][☐] Kalt ist der Abendhauch.m4a
|   |   `-- [✓][☐] Känguru
|   |   |   |   `-- [✓][☐] 01 Die Känguru-Chroniken.m4a
|   |   |   |   |-- [✓][☐] 02 Das Känguru-Manifest.m4a
|   |   |   |   |-- [✓][☐] 03 Die Känguru-Offenbarung.m4a
|   |   |   |   |-- [✓][☐] 04 Die Känguru‐Apokryphen.m4a
|   |   `-- [✓][☐] Katia Fox
|   |   |   |   `-- [✓][☐] 01 Das kupferne Zeichen (1).m4a
|   |   |   |   |-- [✓][☐] 02 Der silberne Falke (2).m4a
|   |   |   |   |-- [✓][☐] 03 Der goldene Thron (3).m4a
|   |   `-- [✓][☑] Kingsbridge
|   |   |   |   `-- [✓][☑] 01 Die Säulen der Erde.m4a
|   |   |   |   |-- [✓][☑] 02 Die Tore der Welt.m4a
|   |   |   |   |-- [✓][☑] 03 Das Fundament der Ewigkeit.m4a
|   |   `-- [✓][☑] Lautlos.m4a
|   |   |-- [✓][☑] Limit.m4a
|   |   |-- [✓][☐] Mein ist die Stunde der Nacht.m4a
|   |   |-- [✓][☐] Meteor.m4a
|   |   |-- [✓][☑] Money - deutsch.m4a
|   |   |-- [✓][☑] Money Master the Game.m4a
|   |   |-- [𐄂][☐] Mordshunger.m4a
|   |   |-- [✓][☐] Nachtschicht.m4a
|   |   |-- [✓][☑] Never - Die letzte Entscheidung.m4a
|   |   |-- [✓][☐] Noah.m4a
|   |   |-- [✓][☑] NSA - Nationales Sicherheits-Amt.m4a
|   |   |-- [✓][☐] Passagier 23.m4a
|   |   `-- [✓][☐] Per Anhalter durch die Galaxis
|   |   |   |   `-- [✓][☐] 01 Per Anhalter durch die Galaxis (1).m4a
|   |   |   |   |-- [✓][☐] 02 Das Restaurant am Ende des Universums (2).m4a
|   |   |   |   |-- [✓][☐] 03 Das Leben, das Universum und der ganze Rest (3).m4a
|   |   |   |   |-- [✓][☐] 04 Macht’s gut, und danke für den Fisch (4).m4a
|   |   |   |   |-- [✓][☐] 05 Einmal Rupert und zurück (5).m4a
|   |   `-- [✓][☑] Pompeji.m4a
|   |   |-- [✓][☐] Puppenmord.m4a
|   |   `-- [✓][☑] Qualityland
|   |   |   |   `-- [✓][☑] 01 Qualityland.m4a
|   |   |   |   |-- [✓][☑] 02 QualityLand 2.0.m4a
|   |   `-- [✓][☐] Reckless
|   |   |   |   `-- [✓][☐] 01 Reckless. Steinernes Fleisch.m4a
|   |   |   |   |-- [✓][☐] 02 Reckless. Lebendige Schatten.m4a
|   |   |   |   |-- [✓][☐] 03 Reckless. Das goldene Garn.m4a
|   |   |   |   |-- [✓][☐] 04 Reckless. Auf silberner Fährte.m4a
|   |   `-- [✓][☐] Robert Langdon
|   |   |   |   `-- [✓][☐] 01 Illuminati.m4a
|   |   |   |   |-- [✓][☐] 02 Sakrileg.m4a
|   |   |   |   |-- [✓][☐] 03 Das verlorene Symbol.m4a
|   |   |   |   |-- [✓][☐] 04 Inferno.m4a
|   |   |   |   |-- [✓][☐] 05 Origin.m4a
|   |   `-- [✓][☐] Schöne neue Welt.m4a
|   |   |-- [✓][☐] Tatort Eifel.m4a
|   |   |-- [✓][☑] Through a Glass Darkly.m4a
|   |   `-- [✓][☐] Tintenwelt
|   |   |   |   `-- [✓][☐] 01 Tintenherz.m4a
|   |   |   |   |-- [✓][☐] 02 Tintenblut.m4a
|   |   |   |   |-- [✓][☐] 03 Tintentod.m4a
|   |   `-- [✓][☐] Tod und Teufel.m4a
|   |   |-- [✓][☐] Tohuwabohu.m4a
|   |   `-- [✓][☑] Warringham
|   |   |   |   `-- [✓][☑] 01 Das Lächeln der Fortuna.m4a
|   |   |   |   |-- [✓][☑] 02 Die Hüter der Rose.m4a
|   |   `-- [✓][☐] Wie man Freunde gewinnt.m4a
|   |   `-- [✓][☑] ZERO.m4a


Remote only items:
A-Gruppe
A-Gruppe/01 Misterioso (1).m4a
A-Gruppe/02 Böses Blut (2).m4a
A-Gruppe/03 Falsche Opfer (3).m4a
A-Gruppe/04 Tiefer Schmerz (4).m4a
A-Gruppe/05 Rosenrot (5).m4a
A-Gruppe/06 Ungeschoren (6).m4a
A-Gruppe/07 Totenmesse (7).m4a
A-Gruppe/08 Dunkelziffer (8).m4a
Die Siedler von Catan.m4a
Die Zucht.m4a
Dirk Gently
Dirk Gently/01 Der elektrische Mönch - Dirk Gently 1.m4a
Dirk Gently/02 Der lange dunkle Fünfuhrtee der Seele - Dirk Gently 2.m4a
Dirk Gently/03 Lachs im Zweifel - Dirk Gently 3.m4a
Dora Gudmundsdottir
Dora Gudmundsdottir/01 Das letzte Ritual.m4a
Dora Gudmundsdottir/02 Das gefrorene Licht.m4a
Dora Gudmundsdottir/03 Das glühende Grab.m4a
Dora Gudmundsdottir/04 Die eisblaue Spur.m4a
Dora Gudmundsdottir/05 Feuernacht.m4a
Dora Gudmundsdottir/06 Todesschiff.m4a
Eberhofer
Eberhofer/01 Winterkartoffelknödel (1).m4a
Eberhofer/02 Dampfnudelblues (2).m4a
Eberhofer/03 Schweinskopf al dente (3).m4a
Eberhofer/04 Grießnockerlaffäre (4).m4a
Eberhofer/05 Sauerkrautkoma (5).m4a
Eberhofer/06 Zwetschgendatschikomplott (6).m4a
Eberhofer/07 Leberkäsjunkie (7).m4a
Eberhofer/08 Weißwurstconnection (8).m4a
Eberhofer/09 Kaiserschmarrndrama (9).m4a
Eberhofer/10 Guglhupfgeschwader (10).m4a
Eberhofer/11 Rehragout-Rendezvous (11).m4a
Eisfieber.m4a
Flieh in die dunkle Nacht.m4a
Geisterfjord.m4a
Hitchhiker's Guide to the Galaxy
Hitchhiker's Guide to the Galaxy/01 Hitchhiker's Guide to the Galaxy (1).m4a
Hitchhiker's Guide to the Galaxy/02 The Restaurant at the End of the Universe (2).m4a
Hitchhiker's Guide to the Galaxy/03 Life the Universe and Everything (3).m4a
Hitchhiker's Guide to the Galaxy/04 So Long and Thanks for All the Fish (4).m4a
Hitchhiker's Guide to the Galaxy/05 Mostly Harmless (5).m4a
Im Wald der stummen Schreie.m4a
In Sachen Signora Brunetti.m4a
Kalt ist der Abendhauch.m4a
Känguru
Känguru/01 Die Känguru-Chroniken.m4a
Känguru/02 Das Känguru-Manifest.m4a
Känguru/03 Die Känguru-Offenbarung.m4a
Känguru/04 Die Känguru‐Apokryphen.m4a
Katia Fox
Katia Fox/01 Das kupferne Zeichen (1).m4a
Katia Fox/02 Der silberne Falke (2).m4a
Katia Fox/03 Der goldene Thron (3).m4a
Mein ist die Stunde der Nacht.m4a
Meteor.m4a
Nachtschicht.m4a
Noah.m4a
Passagier 23.m4a
Per Anhalter durch die Galaxis
Per Anhalter durch die Galaxis/01 Per Anhalter durch die Galaxis (1).m4a
Per Anhalter durch die Galaxis/02 Das Restaurant am Ende des Universums (2).m4a
Per Anhalter durch die Galaxis/03 Das Leben, das Universum und der ganze Rest (3).m4a
Per Anhalter durch die Galaxis/04 Macht’s gut, und danke für den Fisch (4).m4a
Per Anhalter durch die Galaxis/05 Einmal Rupert und zurück (5).m4a
Puppenmord.m4a
Reckless
Reckless/01 Reckless. Steinernes Fleisch.m4a
Reckless/02 Reckless. Lebendige Schatten.m4a
Reckless/03 Reckless. Das goldene Garn.m4a
Reckless/04 Reckless. Auf silberner Fährte.m4a
Robert Langdon
Robert Langdon/01 Illuminati.m4a
Robert Langdon/02 Sakrileg.m4a
Robert Langdon/03 Das verlorene Symbol.m4a
Robert Langdon/04 Inferno.m4a
Robert Langdon/05 Origin.m4a
Schöne neue Welt.m4a
Tatort Eifel.m4a
Tintenwelt
Tintenwelt/01 Tintenherz.m4a
Tintenwelt/02 Tintenblut.m4a
Tintenwelt/03 Tintentod.m4a
Tod und Teufel.m4a
Tohuwabohu.m4a
Wie man Freunde gewinnt.m4a


Profile: ***@***.***
Queued jobs count: 32
[move] Die Siedler von Catan.m4a
[move] Die Zucht.m4a
[move] Eisfieber.m4a
[move] Flieh in die dunkle Nacht.m4a
[move] Geisterfjord.m4a
[move] Im Wald der stummen Schreie.m4a
[move] In Sachen Signora Brunetti.m4a
[move] Kalt ist der Abendhauch.m4a
[move] Mein ist die Stunde der Nacht.m4a
[move] Meteor.m4a
[move] Mordshunger.m4a
[move] Nachtschicht.m4a
[move] Noah.m4a
[move] Passagier 23.m4a
[move] Puppenmord.m4a
[move] Schöne neue Welt.m4a
[move] Tatort Eifel.m4a
[move] Tod und Teufel.m4a
[move] Tohuwabohu.m4a
[move] Wie man Freunde gewinnt.m4a
[move] A-Gruppe
[move] Dirk Gently
[move] Reckless
[move] Tintenwelt
[move] Per Anhalter durch die Galaxis
[move] Katia Fox
[move] Dora Gudmundsdottir
[move] Eberhofer
[move] Hitchhiker's Guide to the Galaxy
[move] Känguru
[move] Robert Langdon
[update] Audiobooks

**Comment by fs1209** [comment]
You’re right, this is the content of the downloaded file:
￼
----------
Programming today is a race between software engineers striving to build
bigger and better idiot-proof programs, and the Universe trying to produce
bigger and better idiots. So far, the Universe is winning.
--Rich Cook

**Comment by GianniCarlo** [comment]
@fs1209 following up on this, I can confirm that most books are already uploaded, and they were just already moved in S3 under the Audiobooks folder, this could be related to an open issue with the third party library (https://github.com/lucas34/SwiftQueue/issues/375) that we'll be replacing with our own implementation, where the task may have already ran successfully, but on the next app launch, it may be queued up again, and that's why it's stuck on moving the file `Die Siedler von Catan.m4a` to the `Audibooks` folder.

I went through the entire library, and the hierarchy of folder matches what we have, we're just missing the following files:

- Mordshunger.m4a
- 05 Rosenrot (5).m4a
- 02 Das gefrorene Licht.m4a
- 04 Goblet of Fire.m4a
- 01 Das Lächeln der Fortuna.m4a

If you sign out, and the sign back in on your phone, these should be queued up, and after that it should sync without problems (another improvement that I'll be working on is background tasks, but after finishing the queue fix, not sure how that got past me 🤦)

Once those are uploaded, on your other devices, you can offload the error file that got associated with your book (on Options → Remove from device), and then redownload and it should work. Heads-up that streaming should work, there's a cache improvement that I have to work on, but got pushed after the queue fix, as that's critical right now

Please let me know how it goes @fs1209

**Comment by fs1209** [comment]
Good Morning!

I had some tries from my iPhone and there was always the point where my iPhone was locked and the upload was interrupted.
Then I did the following:
1. connected all of my devices to my account
2. removed all Books from one of the devices, the empty library synced to all the devices.
3. logged off my account from all devices
4. copied my audiobooks to all my devices
5. logged into my account on my Mac (this syncs perfectly on the first try
6. after upload was done, I’ve logged into all my devices and everything was in sync.

One issue seems to be there when I start listening to a book that’s not downloaded. There are no chapter marks, just one big chapter.

Shall I open an issue on GitHub for this?

Thanks for your support!
Frank

----------
Einfachheit ist die höchste Stufe der Vollendung.
--Leonardo da Vinci

**Comment by GianniCarlo** [comment]
Hi @fs1209 !

Yeah regarding the iPhone locked, the next thing I'm working on is adding background tasks, I thought that it would suffice with just the background time we have while the user is doing playback with the phone locked, but it's definitely not the case. 

I've submitted v5.2.0 that should address the two items I mentioned before in this ticket (preserve order of queued tasks between app launches, and avoid duplicated queued tasks per the ticket open on the third party library). Two other things pending from this ticket is to avoid storing the S3 error response, and give some feedback to the user when an item is being uploaded, so we can show an alert if the user tries to offload the item's backing file

Regarding chapter marks, yeah let's keep track of it in a separate issue. Basically after loading the file's metadata prior to streaming, we reload the chapter info we have on the local database, so we can show the proper chapter marks, so something should be going wrong in that process

**Comment by GianniCarlo** [comment]
An update on this ticket, we should be now giving feedback if the user attempts to delete an item that is being uploaded. The next thing to be worked on is the S3 error response being stored, I'll include the fix in the beta (v5.2.6) later this week

**Comment by GianniCarlo** [comment]
The S3 error response fix is live now on the AppStore, @fs1209 let me know if there's any pending items please

**Comment by fs1209** [comment]
Everything works fine, many thanks for your great support!
